include/aikido/planner/ompl/detail/OMPLConfigurationToConfigurationPlanner-impl.hpp

```diff
@@ -0,0 +1,200 @@
+  , mBoundsProjector(std::move(boundsProjector))
+  , mMaxDistanceBtwValidityChecks(maxDistanceBtwValidityChecks)
+{
+  if (!mInterpolator)
+    mInterpolator = std::make_shared<aikido::statespace::GeodesicInterpolator>(
+        mStateSpace);
+
+  if (!dmetric)
+    dmetric = std::move(createDistanceMetric(mStateSpace));
+
+  // TODO (avk): The constraint namespace functions only work with
+  // metaskeletonstatespace
+  const auto metaskeletonStatespace
+      = std::dynamic_pointer_cast<const MetaSkeletonStateSpace>(mStateSpace);
+
+  if (!metaskeletonStatespace
+      && (!sampler || !boundsConstraint || !boundsProjector))
+  {
+    throw std::invalid_argument(
+        "[OMPLPlanner] Either statespace must be metaSkeletonStateSpace or all "
+        "of sampler, boundsContraint and boundsProjector must be provided.");
+  }
+
+  if (!sampler)
+  {
+    if (!rng)
+      throw std::invalid_argument(
+          "[OMPLPlanner] Both of sampler and rng cannot be nullptr.");
+    sampler = std::move(
+        createSampleableBounds(metaskeletonStatespace, rng->clone()));
+  }
+
+  if (!mBoundsConstraint)
+    mBoundsConstraint = std::move(createTestableBounds(metaskeletonStatespace));
+
+  if (!mBoundsProjector)
+    mBoundsProjector
+        = std::move(createProjectableBounds(metaskeletonStatespace));
+
+  // Geometric State space
+  auto sspace = ompl_make_shared<GeometricStateSpace>(
+      mStateSpace,
+      std::make_shared<aikido::statespace::GeodesicInterpolator>(mStateSpace),
+      std::move(dmetric),
+      std::move(sampler),
+      mBoundsConstraint,
+      std::move(mBoundsProjector));
+
+  // Space Information
+  auto si = ompl_make_shared<::ompl::base::SpaceInformation>(std::move(sspace));
+
+  // Setup the planner
+  mPlanner = ompl_make_shared<PlannerType>(si);
+}
+
+//==============================================================================
+template <class PlannerType>
+trajectory::TrajectoryPtr
+OMPLConfigurationToConfigurationPlanner<PlannerType>::plan(
+    const SolvableProblem& problem, Result* result)
+{
+  auto si = mPlanner->getSpaceInformation();
+
+  // Set validity checker
+  std::vector<constraint::ConstTestablePtr> constraints{problem.getConstraint(),
+                                                        mBoundsConstraint};
+  auto conjunctionConstraint
+      = std::make_shared<constraint::TestableIntersection>(
+          mStateSpace, std::move(constraints));
+  ::ompl::base::StateValidityCheckerPtr vchecker
+      = ompl_make_shared<StateValidityChecker>(si, conjunctionConstraint);
+  si->setStateValidityChecker(vchecker);
+
+  ::ompl::base::MotionValidatorPtr mvalidator
+      = ompl_make_shared<MotionValidator>(si, mMaxDistanceBtwValidityChecks);
+  si->setMotionValidator(mvalidator);
+
+  // Define the OMPL problem
+  auto pdef = ompl_make_shared<::ompl::base::ProblemDefinition>(si);
+  auto sspace
+      = ompl_static_pointer_cast<GeometricStateSpace>(si->getStateSpace());
```

Comment:
Nit: Will the type of `sspace` always be `GeometricStateSpace`? If so, let's add an assertion here for sure something like `assert(ompl_dynamic_pointer_cast<GeometricStateSpace>(si->getStateSpace()))`;